security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -567,12 +585,107 @@ func (sc *SecretCache) generateGatewaySecret(token string, connKey ConnKey, t ti
 	}, nil
 }
 
+// If there is an existing root certificate under a well known path, return true.
+// Otherwise, return false.
+func (sc *SecretCache) rootCertificateExists() bool {
+	b, err := ioutil.ReadFile(existingRootCertFile)
+	if err != nil || len(b) == 0 {
+		return false
+	}
+	return true
+}
+
+// If there is an existing private key and certificate under a well known path, return true.
+// Otherwise, return false.
+func (sc *SecretCache) keyCertificateExists() bool {
+	b, err := ioutil.ReadFile(existingCertChainFile)
+	if err != nil || len(b) == 0 {
+		return false
+	}
+	b, err = ioutil.ReadFile(existingKeyFile)
+	if err != nil || len(b) == 0 {
+		return false
+	}
+
+	return true
+}
+
+// Generate a root certificate item from the existing root certificate file
+// under a well known path.
+func (sc *SecretCache) generateRootCertFromExistingFile(token string, connKey ConnKey) (*model.SecretItem, error) {
+	rootCert, err := ioutil.ReadFile(existingRootCertFile)
+	if err != nil {
+		return nil, err
+	}
+
+	t := time.Now()
+	var certExpireTime time.Time
+	if certExpireTime, err = nodeagentutil.ParseCertAndGetExpiryTimestamp(rootCert); err != nil {
+		cacheLog.Errorf("failed to extract expire time from the root certificate: %v", err)
+		return nil, fmt.Errorf("failed to extract expire time from the root certificate: %v", err)
+	}
+
+	// Set the rootCert
+	sc.rootCertMutex.Lock()
+	sc.rootCert = rootCert
+	sc.rootCertMutex.Unlock()
+
+	return &model.SecretItem{
+		ResourceName: connKey.ResourceName,
+		RootCert:     rootCert,
+		ExpireTime:   certExpireTime,
+		Token:        token,
+		CreatedTime:  t,
+		Version:      t.String(),
+	}, nil
+}
+
+// Generate a key and certificate item from the existing key certificate files
+// under a well known path.
+func (sc *SecretCache) generateKeyCertFromExistingFiles(token string, connKey ConnKey) (*model.SecretItem, error) {
```

Comment:
generateWorkloadKeyCertFromExistingFiles

Reply:
The file based SDS may be used on an ingress gateway also.